README.txt
This is an implementation of the Majority Judgement voting procedure, as
proposed by Michel Balinski and Rida Laraki in `A theory of measuring, electing
, and ranking <http://www.pnas.org/content/104/21/8720.full>`_

It takes tallies of grades provided by voters and provides a python type to 
wrap them which is ordered according to the majority judgement order. 

For example::

    x = MajorityJudgement([5, 5])
    y = MajorityJudgement([3, 7])
    assert x < y

The first is the grade for a candidate who has received 5 votes of 0 and 5 votes
of 1. The second is a candidate with 3 votes of 0 and 7 of 1. As a result the 
majority judgement procedure ranks the first candidate as being worse than the 
second candidate.

In the event that you want more information about the details of the voting,
the MajorityJudgement objects behave like the corresponding lists of grades
assigned to them.

So for example::

    assert list(MajorityJudgement([2,2])) == [0,1,0,1]

The MajorityJudgement object can generally be used exactly like its 
corresponding list of grades. In particular it supports all the container 
methods in an efficient manner.

Internally the implementation is rather different. It is encoded as a list of 
repeating cycles. So the list

    [0, 0, 0, 1, 0, 1, 0, 1]

is internally represented as 

    [[[0], 3], [[1], 1], [[0, 1], 2]]

The representation we choose only ever builds cycles of length one or two: A 
cycle of length one is usually built when there is an unambiguous median, while
a cycle of length 2 will usually be built when the lower and upper medians are 
different (this isn't entirely true. e.g. the list of votes [1,1,1,1] will not
produce any cycles of length 2, but it's a good rule of thumb)

Rather than building the whole list and then compressing, entire cycles are 
built at once by working how large a run can be popped from the remaining votes
and popping it all at once. This means that even very large populations of 
voters may be efficiently worked with as the resulting lists are actually very
small.

The comparisons too may be performed efficiently on this compressed
representation: Equality works straightforwardly in the sense that two equal
votes will produce two equal representations. Comparisons may be performed by 
popping common prefixes first: So if two lists both start with [x, n] and [x,m]
then we may throw away min(m,n) entries immediately rather than having to 
iterate through.

It appears to be the case that all votes can be represented as a compressed 
list of this form with very few cycles. I conjecture that it can be done in 
O(log(voters)), but have not verified this. Certainly in practice it seems to 
be the case.


CHANGES.txt
v0.0.1, 17/02/2013 -- Initial release
v0.0.2, 17/02/2013 -- Trivial packaging problems
v0.1.0, 17/02/2013 -- Switch to more sophisticated algorithm that better run length encodes the results
v0.1.1, 18/02/2013 -- Significant improvement of behaviour in the presence of cycles of length 2.
v0.1.2, 18/02/2013 -- Improved compression of the resulting list. Make behaviour more collection like.
v0.2.0, 20/02/2013 -- Implementation no longer lazily evaluates judgement trail as they're always really short. Simplified code.
v0.3.0, 20/02/2013 -- Greatly stripped down API - just implement the basic comparison operations. Also significantly improved implementation.


majority_judgement_test.py
from majority_judgement import MajorityJudgement
import pytest
import re


class TestMajorityJudgementSoundness:
    example_votes = [
        [1, 1, 1, 1, 1, 1, 1, 1],
        [1, 2, 3],
        [13],
        [1, 13, 12, 11, 1],
        [1, 1, 1, 1],
        [2, 3],
        [2, 1, 3],
        [10, 1, 10],
        [1, 3, 2],
        [10],
        [9],
        [10],
        [9, 1]
    ]

    on_all_examples = pytest.mark.parametrize("x", example_votes)

    @on_all_examples
    def test_tallies(self, x):
        tallies = [0 for _ in x]
        for i in MajorityJudgement(x):
            tallies[i] = tallies[i] + 1
        assert x == tallies

    @on_all_examples
    def test_not_less_than_self(self, x):
        assert MajorityJudgement(x) >= MajorityJudgement(x)

    @on_all_examples
    def test_front_loading_moves_earlier(self, x):
        for i in xrange(1, len(x)):
            y = list(x)
            y[i] = y[i] - 1
            y[i-1] = y[i-1] + 1
            assert MajorityJudgement(y) < MajorityJudgement(x)

    @on_all_examples
    def test_equality_of_same_input(self, x):
        x = MajorityJudgement(x)
        assert x == x

    @on_all_examples
    def test_equals_self_when_fully_evaluated(self, x):
        x1 = MajorityJudgement(x)
        x2 = MajorityJudgement(x)
        list(x1)
        assert x1 == x2

    @on_all_examples
    def test_calling_list_twice_produces_same_result(self, x):
        x = MajorityJudgement(x)
        assert list(x) == list(x)

    @on_all_examples
    def test_equality_of_identical_inputs(self, x):
        assert MajorityJudgement(x) == MajorityJudgement(x)

    @pytest.mark.parametrize(("x", "y"), [
        (x, y) for x in example_votes for y in example_votes if x != y
    ])
    def test_inequality_of_different_inputs(self, x, y):
        assert MajorityJudgement(x) != MajorityJudgement(y)

    @pytest.mark.parametrize(("x", "y"), [
        (x, y) for x in example_votes for y in example_votes
    ])
    def test_anti_symmetry_of__compare(self, x, y):
        assert (MajorityJudgement(x)._compare(MajorityJudgement(y)) ==
                -MajorityJudgement(y)._compare(MajorityJudgement(x)))

    @on_all_examples
    def test_length_agrees_with_list(self, x):
        assert len(MajorityJudgement(x)) == len(list(MajorityJudgement(x)))

    @on_all_examples
    def test_length_agrees_with_list_when_evaluated(self, x):
        x = MajorityJudgement(x)
        y = list(x)
        assert len(x) == len(y)

    @on_all_examples
    def test_correct_number(self, x):
        total = sum(x)
        assert len(list(MajorityJudgement(x))) == total

    @on_all_examples
    def test_positive_index_matches_list_index(self, x):
        y = list(MajorityJudgement(x))
        x = MajorityJudgement(x)

        for i in xrange(len(y)):
            assert x[i] == y[i]

    @on_all_examples
    def test_negative_index_matches_list_index(self, x):
        y = list(MajorityJudgement(x))
        x = MajorityJudgement(x)

        for i in xrange(len(y)):
            assert x[-i] == y[-i]

    @pytest.mark.parametrize(("x", "m"), [
        ([1, 1, 1, 1], 1),
        ([5, 1, 1, 5], 1),
        ([5, 1, 1, 6], 2),
        ([1, 1, 1, 1, 20], 4)
    ])
    def test_starts_with_median(self, x, m):
        assert MajorityJudgement(x)[0] == m

    @pytest.mark.parametrize(
        ("x", "y", "k"),
        [(x, y, k)
            for x in example_votes
            for y in example_votes
            if sum(x) == sum(y) and x != y
            for k in xrange(sum(x) / 2)])
    def test_can_drop_outliers(self, x, y, k):
        assert sum(x) == sum(y)
        x1 = MajorityJudgement(x)
        y1 = MajorityJudgement(y)

        def drop_outliers(l, k):
            l = list(l)
            for r in [xrange(len(l)), reversed(xrange(len(l)))]:
                k2 = k
                for i in r:
                    if l[i] >= k2:
                        l[i] = l[i] - k2
                        break
                    else:
                        k2 = k2 - l[i]
                        l[i] = 0
            return l

        x2 = MajorityJudgement(drop_outliers(x, k))
        y2 = MajorityJudgement(drop_outliers(y, k))

        if x1 <= y1:
            assert x2 <= y2
        else:
            assert y2 <= x2

    def build_single_upvote(self, k, n):
        return [k] + [0 for _ in xrange(n-1)] + [1]

    @pytest.mark.parametrize(("k", "n"), [
        (10, 1),
        (10, 5),
        (10, 3),
        (1, 1),
        (2, 10)
    ])
    def test_single_upvote_comes_last(self, k, n):
        x = self.build_single_upvote(k, n)
        l = list(MajorityJudgement(x))
        assert l == [0 for _ in xrange(k)] + [n]

    @pytest.mark.parametrize(("k", "m", "n"), [
        (10, 1, 5),
        (10, 5, 6),
        (10, 3, 5),
        (1, 1, 2),
        (2, 10, 20)
    ])
    def test_single_upvote_determines_result(self, k, m, n):
        assert (MajorityJudgement(self.build_single_upvote(k, n)) >
                MajorityJudgement(self.build_single_upvote(k, m)))

    order_tests = pytest.mark.parametrize(("x", "y"), [
        ([10], [5, 5]),
        ([10], [9, 1]),
        ([9, 1], [5, 5]),
        ([1, 1, 1, 1, 1], [0, 1, 1, 1, 2]),
        ([5, 5], [10, 10])
    ])

    @order_tests
    def test_on_specified_pairs_unevaluated(self, x, y):
        assert MajorityJudgement(x) < MajorityJudgement(y)
        assert MajorityJudgement(y) > MajorityJudgement(x)

    @order_tests
    def test_on_specified_pairs_fully_evaluated(self, x, y):
        xe = MajorityJudgement(x)
        ye = MajorityJudgement(y)
        list(xe)
        list(ye)
        assert xe < ye
        assert ye > xe

    @pytest.mark.parametrize(("ev"), [example_votes])
    def test_sorts_like_corresponding_lists(self, ev):
        by_mj = [list(x) for x in sorted([MajorityJudgement(y) for y in ev])]
        by_list = sorted([list(MajorityJudgement(y)) for y in ev])
        assert by_mj == by_list

    @on_all_examples
    def test_repr_does_not_evaluate(self, x):
        x = MajorityJudgement(x)
        repr(x)
        assert len(x._judgement_trail) == 0

    @on_all_examples
    def test_repr_reflects_evaluation(self, x):
        x = MajorityJudgement(x)
        o = repr(x)
        list(x)
        assert repr(x) != o

    def test_empty_behaviour(self):
        assert MajorityJudgement([]) == MajorityJudgement([])
        assert MajorityJudgement([]) < MajorityJudgement([1])
        assert MajorityJudgement([1]) > MajorityJudgement([])

    def test_no_delete(self):
        x = MajorityJudgement([1, 2, 3])
        l = list(x)
        with pytest.raises(TypeError) as excinfo:
            del x[0]
        assert re.search('not support modifying', excinfo.value.message)
        assert list(x) == l

    def test_no_set(self):
        x = MajorityJudgement([1, 2, 3])
        l = list(x)
        with pytest.raises(TypeError) as excinfo:
            x[0] = 15
        assert re.search('not support modifying', excinfo.value.message)
        assert list(x) == l

    def test_index_out_of_bounds(self):
        with pytest.raises(IndexError):
            MajorityJudgement([1, 2, 3])[6]
        with pytest.raises(IndexError):
            MajorityJudgement([1, 2, 3])[-6]


majorityjudgement/test/test_compare_rle.py
from majorityjudgement.compare_rle import compare as compare_rle
import pytest

def run_length_compress_list(xs):
    result = []
    last_x = None
    for x in xs:
        if x == last_x:
            l = result[-1]
            l[1] = l[1] + 1
        else:
            result.append([x, 1])
            last_x = x
    return result

@pytest.mark.parametrize(("x", "y"), [
    ([1,1,1,1,5], [1,1,1,1]),
    ([1,1,1,1,5], [1,1,1,1,4]),
    ([1,2,1,2,1], [1,2,2,1,2])
])
def test_run_length_preserves_order(x, y):
    xr = iter(run_length_compress_list(x))
    yr = iter(run_length_compress_list(y))

    if x < y:
        assert compare_rle(xr, yr) < 0
    else:
        assert compare_rle(xr, yr) > 0


majorityjudgement/test/test_cyclicgenerator.py
from majorityjudgement.cyclicgenerator import CyclicGenerator
import pytest

@pytest.mark.parametrize("x", [
    [([1,2,3], 5)],
    [([1], 10)],
    [([1,2,3], 5), ([1], 10)],
    [([1], 10), ([1], 5)],
    [([1,2,3], 5), ([1,2,3], 5)],
])
def test_matches_list(x):
    list_directly = []
    for y, n in x:
        list_directly += y * n
    list_via_generator = []
    for y, n in CyclicGenerator(iter(x)):
        list_via_generator += [y] * n

    assert list_directly == list_via_generator


majority_judgement.py
"""
Majority Judgement is a method of voting proposed by Michel Balinski 
and Rida Laraki in "A theory of measuring, electing, and ranking"
(http://www.pnas.org/content/104/21/8720.full)

Members of the electorate vote by assigning grades to candidates. These grades 
can be any ordinal values (numbers, "A,B,C", "Good, Bad,Terrible", etc. 
All that matters is the relative ordinal positions of the grades.

In this module we assume that grades are consecutive integer values starting
from 0 as the lowest. Because all that matters is the ordering, any other 
grading scheme can be trivially converted to this form.

The essential idea of majority judgement is that we sort the grades assigned to
the candidate in order of most significant to least significant. At any given
point the most significant grade of those left is the lower median of the set
(i.e. the highest grade which at least 50% of the population supports).

So for example given the grading Bad, OK, OK, Good we would convert this to the
sequence

OK, Bad, OK, Good

Another candidate might have the grading

OK, Good, OK, Good

This candidate would win because their second grading is better than the first
candidate's.

This module provides a type which wraps a tally of grades and is then ordered
in terms of the majority judgement. It may then be used to implement a voting
procedure by assigning each candidate their tally and taking the maximum.
"""

class MajorityJudgement:
    """
    Objects of type MajorityJudgement behave like a lazily evaluated frozen 
    list. They may be indexed, iterated over and compared exactly as if they
    were their list of majority judgement grades.
    """
    def __init__(self, votes):
        """
        Create a MajorityJudgement object from a tally of grades. Note that
        the votes are taken as tallies, not as a list of grades. i.e. 
        [1,2,1] means that there is one vote each of grades 0 and 2 and 2 votes
        of grade 1, not that there 2 votes of grade 1 and 1 of grade 2.
        """
        self.length = sum(votes)
        self.votes = list(votes)
        self.votes_remaining = sum(votes)
        self.judgement_trail = []

    def __repr__(self):
        return "MajorityJudgement(%s, %s)" % (self.judgement_trail, self.votes)

    def __eq__(self, other):
        return self.compare(other) == 0

    def __ne__(self, other):
        return self.compare(other) != 0

    def __lt__(self, other):
        return self.compare(other) < 0

    def __le__(self, other):
        return self.compare(other) <= 0

    def __gt__(self, other):
        return self.compare(other) > 0

    def __ge__(self, other):
        return self.compare(other) >= 0

    def __len__(self):
        return self.length

    def __getitem__(self, i):
        l = len(self)
        if i < 0 and i > -l:
            i = i + l
        elif i < 0 or i >= l:
            raise IndexError("Index %d out of range [0, %d)", i, len(self))
        for x, n in self.each_judgement():    # pragma: no branch
            if i < n:
                return x
            i = i - n

    def __delitem__(self, i):
        raise TypeError(
            "MajorityJudgement objects do not support modifying"
            "the contents")

    def __setitem__(self, i, x):
        raise TypeError(
            "MajorityJudgement objects do not support modifying"
            "the contents")

    def __iter__(self):
        for (x, n) in self.each_judgement():
            for _ in xrange(n):
                yield x

    def compare(self, other):
        """
            Return an integer expressing the order relation between self and 
            other. -1 if self < other, 0 if self == other, 1 if self > other.

            Ordering is defined to be equivalent to lexicographical ordering of
            list(self) and list(other), but may not fully evaluate the sequence
            and may require less iteration due to long stretches of common 
            values.
        """
        if self is other:
            return 0
        if len(self) == 0 and len(other) == 0:
            return 0
        if len(self) == 0:
            return -1
        if len(other) == 0:
            return 1

        si = self.each_judgement()
        oi = other.each_judgement()

        x, xn = si.next()
        y, yn = oi.next()

        while True:
            if x < y:
                return -1
            if y < x:
                return 1

            m = min(xn, yn)
            xn = xn - m
            yn = yn - m

            if xn == 0:
                try:
                    x, xn = si.next()
                except StopIteration:
                    if yn > 0:
                        return -1
                    else:
                        try:
                            oi.next()
                            return -1
                        except StopIteration:
                            return 0

            if yn == 0:
                try:
                    y, yn = oi.next()
                except StopIteration:
                    # The fact that we've got this far means that xn > 0 so
                    # there is remaining x
                    return 1

    def each_judgement(self):
        for x in self.judgement_trail:
            yield x
        while len(self.votes) > 0:
            tot = 0
            for i in xrange(len(self.votes)):  # pragma: no branch
                tot += self.votes[i]
                if 2 * tot >= self.votes_remaining:
                    votes_to_pop = 1
                    self.votes_remaining -= votes_to_pop
                    self.votes[i] -= votes_to_pop

                    while len(self.votes) > 0 and self.votes[-1] <= 0:
                        self.votes.pop()

                    if len(self.judgement_trail) > 0:
                        xv = self.judgement_trail[-1]
                    else:
                        xv = None

                    if xv and xv[0] == i:
                        xv[1] = xv[1] + votes_to_pop
                    else:
                        self.judgement_trail.append([i, votes_to_pop])

                    yield [i, 1]
                    break


majorityjudgement/cyclicgenerator.py
class CyclicGenerator:

    def __init__(self, generator):
        self.generator = generator
        self.current_cycle = None
        self.current_index = 0
        self.current_count = 0

    def next(self):
        if not self.current_cycle or self.current_count <= 0:
            (self.current_cycle, self.current_count) = self.generator.next()
            self.current_index = 0
        if len(self.current_cycle) == 1:
            result = (self.current_cycle[0], self.current_count)
            self.current_cycle = None
            return result
        if self.current_index >= len(self.current_cycle):
            self.current_index = 0
            self.current_count = self.current_count - 1
        if self.current_count <= 0:
            return self.next()
        result = self.current_cycle[self.current_index]
        self.current_index = self.current_index + 1
        return (result, 1)

    def __iter__(self):
        return self


majorityjudgement/grading.py
"""
Majority Judgement is a method of voting proposed by Michel Balinski
and Rida Laraki in "A theory of measuring, electing, and ranking"
(http://www.pnas.org/content/104/21/8720.full)

Members of the electorate vote by assigning grades to candidates. These grades
can be any ordinal values (numbers, "A,B,C", "Good, Bad,Terrible", etc.
All that matters is the relative ordinal positions of the grades.

In this module we assume that grades are consecutive integer values starting
from 0 as the lowest. Because all that matters is the ordering, any other
grading scheme can be trivially converted to this form.

The essential idea of majority judgement is that we sort the grades assigned to
the candidate in order of most significant to least significant. At any given
point the most significant grade of those left is the lower median of the set
(i.e. the highest grade which at least 50% of the population supports).

So for example given the grading Bad, OK, OK, Good we would convert this to the
sequence

OK, Bad, OK, Good

Another candidate might have the grading

OK, Good, OK, Good

This candidate would win because their second grading is better than the first
candidate's.

This module provides a type which wraps a tally of grades and is then ordered
in terms of the majority judgement. It may then be used to implement a voting
procedure by assigning each candidate their tally and taking the maximum.
"""

import collections
import copy


class MajorityJudgement():
    """
    Objects of type MajorityJudgement support comparison and ordering options 
    as per the ordering of the described voting algorithm. They expose no other
    operations.
    """
    def __init__(self, tally):
        """
        Create a MajorityJudgement object from a tally of grades. Note that
        the votes are taken as tallies, not as a list of grades. i.e.
        [1,2,1] means that there is one vote each of grades 0 and 2 and 2 votes
        of grade 1, not that there 2 votes of grade 1 and 1 of grade 2.
        """

        for x in tally:
            if type(x) is not int:
                raise ValueError("Tally counts must be integers: %s" % tally)
            if x < 0:
                raise ValueError("Tally counts may not be negative: %s" % tally)
        
        self.tally = tally
        self.__judgement_trail = []
        self.__hangover = -1
        self.__calculate_judgement_trail()

    def __repr__(self):
        return "MajorityJudgement(tally=%s,judgement_trail=%s)" % (self.tally, self.__judgement_trail)

    def __eq__(self, other):
        return self.__judgement_trail == other.__judgement_trail

    def __ne__(self, other):
        return self.__judgement_trail != other.__judgement_trail

    def __lt__(self, other):
        return self._compare(other) < 0

    def __le__(self, other):
        return self._compare(other) <= 0

    def __gt__(self, other):
        return self._compare(other) > 0

    def __ge__(self, other):
        return self._compare(other) >= 0

    def _compare(self, other):
        """
        Return an integer expressing the order relation between self and
        other. -1 if self < other, 0 if self == other, 1 if self > other.

        The ordering relationship is defined by the majority judgement 
        voting order
        """
        if self is other:
            return 0
        if not self.__judgement_trail and not other.__judgement_trail:
            return 0
        if not self.__judgement_trail:
            return -1
        if not other.__judgement_trail:
            return 1

        self_stack = list(self.__judgement_trail)
        other_stack = list(other.__judgement_trail)

        si = 0
        oi = 0
        
        while si < len(self_stack) and oi < len(other_stack):
            if self_stack[si] < other_stack[oi]: return -1
            elif self_stack[si] > other_stack[oi]: return 1

            if self_stack[si+1] < other_stack[oi+1]: return -1
            elif self_stack[si+1] > other_stack[oi+1]: return 1

            m = min(self_stack[si+2], other_stack[oi+2])
            self_stack[si+2] -= m
            other_stack[oi+2] -= m

            if self_stack[si+2] == 0: si += 3
            if other_stack[oi+2] == 0: oi += 3

        if si < len(self_stack): return 1
        if oi < len(other_stack): return -1

        return 0

    def __how_many_to_pop(self, preceding, votes, total):
        return 1 + min(total - 2 * preceding - 1,
                       2 * preceding + 2 * votes - total)
    
    def __calculate_judgement_trail(self):
        tallies = [2 * x for x in self.tally]
        tallies_remaining = sum(tallies)
        while len(tallies) > 0:
            tot = 0
            for i in xrange(len(tallies)):  # pragma: no branch
                preceding = tot
                v = tallies[i]
                tot += v
                if 2 * tot >= tallies_remaining:
                    if 2 * tot == tallies_remaining:
                        next_index = i+1
                        while not tallies[next_index]:
                            next_index += 1

                        relevant_indices = [i, next_index]
                        votes_to_pop = self.__how_many_to_pop(preceding,
                                                              v + tallies[next_index],
                                                              tallies_remaining)
                        k = votes_to_pop / 2
                        votes_to_pop = k * 2
                    else:
                        relevant_indices = [i]
                        votes_to_pop = self.__how_many_to_pop(preceding,
                                                              v,
                                                              tallies_remaining)
                        k = votes_to_pop

                    tallies_remaining -= votes_to_pop
                    for i in relevant_indices:
                        tallies[i] -= k

                    while len(tallies) > 0 and tallies[-1] <= 0:
                        tallies.pop()
                    self.__append(relevant_indices, k)
                    break

        assert self.__hangover < 0
        self.__judgement_trail = tuple(self.__judgement_trail)

    def __append(self,xs, n):
        if self.__hangover < 0:
            if len(xs) == 2:
                self.__simple_append(xs[0],xs[1],n)
            else:
                x = xs[0]
                if n > 1: self.__simple_append(x,x,n/2)
                if n % 2 != 0: self.__hangover = x
        else:
            # I actually haven't reasoned through this bit yet.
            # I think we can never have len(xs) == 2 here when the original
            # total passed in is even. Certainly this seems to be true 
            # empirically
            assert len(xs) == 1
            x = xs[0]
            self.__simple_append(self.__hangover, x, 1)
            n -= 1 
            if n > 1: self.__simple_append(x,x,n/2)
            if n % 2 != 0: self.__hangover = x
            else: self.__hangover = -1

    def __simple_append(self,x,y,n):
        self.__judgement_trail.append(x)
        self.__judgement_trail.append(y)
        self.__judgement_trail.append(n)
        
